src/core/packet.lua

```diff
@@ -99,6 +99,12 @@ function data (p) return p.data end
 -- Return packet data length.
 function length (p) return p.length end
 
+-- Set packet data length.
+function resize (p, len)
+   assert(len <= max_payload, "packet payload overflow")
+   p.length = len
+end
+
```

Comment:
This `resize` function seems potentially error-prone to me: if you resize the packet but don't fill in all of the data then you will be leaking unknown information from the previous user of the packet.

One alternative would be for encapsulation/encryption to use a new `packet.pad(p, n)` that always fills with zeros and for decapsulation/decryption to use a new function akin to `packet.from_pointer()` to set payload and size at the same time.